// @marko/tags-api-preview/dist/components/lifecycle/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var import_patch_lifecycle = __toESM(require("../../util/patch-lifecycle"));
const metaKey = Symbol();
const indexKey = Symbol();
const lifecycleMethods = {
  onMount: runMount,
  onUpdate: runUpdate,
  onDestroy: runDestroy
};
function runMount() {
  const meta = this[metaKey];
  if (meta) {
    this[indexKey] = 0;
    for (let i = 1; i < meta.length; i += 2) {
      const handlers = meta[i];
      if (handlers.onMount) {
        handlers.onMount();
      }
    }
  }
}
function runUpdate() {
  const meta = this[metaKey];
  if (meta) {
    this[indexKey] = 0;
    for (let i = 0; i < meta.length; i += 2) {
      if (meta[i]) {
        meta[i] = 0;
        const handlers = meta[i + 1];
        if (handlers.onUpdate) {
          handlers.onUpdate();
        }
      }
    }
  }
}
function runDestroy() {
  const meta = this[metaKey];
  if (meta) {
    for (let i = 1; i < meta.length; i += 2) {
      const handlers = meta[i];
      if (handlers.onDestroy) {
        handlers.onDestroy();
      }
    }
  }
}
module.exports = function(component, cur = {}) {
  const meta = component[metaKey];
  const index = component[indexKey];
  if (meta) {
    if (index === void 0) {
      meta.push(0, cur);
    } else {
      component[indexKey] += 2;
      const prev = meta[index + 1];
      prev.onMount = cur.onMount;
      prev.onDestroy = cur.onDestroy;
      if (prev.onUpdate !== cur.onUpdate) {
        meta[index] = 1;
        prev.onUpdate = cur.onUpdate;
      }
    }
  } else {
    (0, import_patch_lifecycle.default)(component, lifecycleMethods);
    component[metaKey] = [0, cur];
  }
};
